### PR TITLE
BUG, DOC: Allow custom line terminator with delim_whitespace=True

### LINKED ISSUE #12912: DOC, BUG: Clarify and Standardize Whitespace Delimiter Behaviour with Custom Line Terminator
~~~python
>>> from pandas import read_csv
>>> from pandas.compat import StringIO
>>> data = """a b c~1 2 3~4 5 6~7 8 9"""
>>> df = read_csv(StringIO(data), lineterminator='~', delim_whitespace=True)
>>> df
Empty DataFrame
Columns: [a, b, c~1, 2, 3~4, 5, 6~7, 8, 9]
Index: []
~~~
expected:
~~~python
>>> df
    a    b    c
0   1    2    3
1   4    5    6
2   7    8    9
~~~

Note that this bug is only for the C engine, as the Python engine does not yet support `delim_whitespace`.

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.  Closes #12912.

### CODE REVIEW

**Comment by gfyoung** [comment]
The tokenizers in `tokenizer.c`, and the `CParser` tests seem to have a lot of repetition, although reworking the latter seems easier to do than the former.  Can / should this PR also include refactoring changes as well to those parts of the code?

**Comment by jreback** [comment]
no make smaller / simpler PRs that do 1 thing

**Comment by jreback** [comment]
they can of course be built upon one another

File: doc/source/io.rst [hunk 1]
```diff
@@ -97,6 +97,11 @@ sep : str, defaults to ``','`` for :func:`read_csv`, ``\t`` for :func:`read_tabl
   Regex example: ``'\\r\\t'``.
 delimiter : str, default ``None``
   Alternative argument name for sep.
+delim_whitespace : boolean, default False
```
Comment (jreback) [inline doc/source/io.rst]: say this is equivalent to \s+ regex

**Comment by jreback** [comment]
why do you think adding all of this untested code is a good idea?

you would basically have to duplicate all of the test suite with a custom terminator in order to validate this

you need to use functions that are already tested
iow each of the cases needs to all a function

**Comment by gfyoung** [comment]
What functions are you talking about?  If you look at the other tokenizer functions, they all duplicate the same or have very similar case work.  This function in fact is a near duplicate of the `tokenize_whitespace` save replacing checks for `\n` and `\r` with `self->lineterminator`.  That's why I asked about the refactoring.

**Comment by jreback** [comment]
well then refactor first - we can't add more code like this

File: doc/source/io.rst [hunk 2]
```diff
@@ -97,6 +97,11 @@ sep : str, defaults to ``','`` for :func:`read_csv`, ``\t`` for :func:`read_tabl
...
+  Specifies whether or not whitespace (e.g. ``' '`` or ```'\t'``)
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: One backtick too much before `\t`

File: doc/source/io.rst (hunk 1 above)
Comment (gfyoung) [inline doc/source/io.rst]: Done.

File: doc/source/io.rst (hunk 2 above)
Comment (gfyoung) [inline doc/source/io.rst]: Oh, good catch.  Will fix.

**Comment by gfyoung** [comment]
Managed to refactor code so that it only splits based on whether or not whitespace is used as a delimiter, so there are only two tokenizing functions now instead of three (or four after my second commit).

**Comment by jreback** [comment]
@gfyoung ok much better. pls also run asv's for csv to make sure no degredation.

So macros are good because they make the code shorter / more understandable while hopefully not sacrificing perf. more combining code is better (again could be done later), but since you are already working on it....

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -209,6 +209,11 @@
 warn_bad_lines : boolean, default True
     If error_bad_lines is False, and warn_bad_lines is True, a warning for each
     "bad line" will be output. (Only valid with C parser).
+delim_whitespace : boolean, default False
+    Specifies whether or not whitespace (e.g. ``' '`` or ``'\t'``) will be used
+    as the delimiter. Equivalent to ``'\+s'`` in regex. If this option is set
```
Comment (jreback) [inline pandas/io/parsers.py]: say (both her and in the docs), that ``delim_whitespace=True`` equiv to ``sep='\s+'`` (not just that its the same regex)

File: pandas/src/parser/tokenizer.c [hunk 4]
```diff
@@ -1072,169 +1038,177 @@ int tokenize_delim_customterm(parser_t *self, size_t line_limit)
...
+                // start quoted field
                 self->state = IN_QUOTED_FIELD;
-            }
-            else if (c == self->escapechar) {
-                /* possible escaped character */
+            } else if (c == self->escapechar) {
+                // possible escaped character
                 self->state = ESCAPED_CHAR;
-            }
-            else if (c == ' ' && self->skipinitialspace)
```
Comment (jreback) [inline pandas/src/parser/tokenizer.c]: where is this now? (also make a macro for this, e.g. the skipintialspace check)

Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: Split the view on the diff, and you'll see it's still there.  This is one of those moments when Git's diff isn't super human friendly.

**Comment by gfyoung** [comment]
@jreback : any ideas as to that OSX failure?  It doesn't seem related to anything that I changed.

File: pandas/io/parsers.py (hunk 3 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

**Comment by gfyoung** [comment]
@jreback : there seems to have been a recent update to `python-dateutil` (just today), causing a test failure in `tslib` on OSX.  I can reproduce this failure on Linux too FYI.

**Comment by jreback** [comment]
@gfyoung ok! my pip install catches things then, great! (the osx build will pip install python-dateutil to catch things just like this :>), while the 3.5 build uses conda (which is with a stable build). IIRC this has some reverses from previous things, so will have to sort them and see. Maybe we should ban certain versions of dateutil. These can cause weird things to happen.

**Comment by gfyoung** [comment]
@jreback : is it possible to put in a stopgap measure FTTB?

**Comment by jreback** [comment]
see #12944 (I skipped the test for now on master), so rebase.

**Comment by gfyoung** [comment]
Whoot!  Squashed it down into one function!  Hopefully Travis will give the green light this time around.  I will also do some timing analysis for `read_csv` and post it in the conversation.

**Comment by gfyoung** [comment]
This PR
=======
io_bench.frame_to_csv.time_frame_to_csv           188.25ms
io_bench.frame_to_csv2.time_frame_to_csv2         312.94ms
...formatting.time_frame_to_csv_date_formatting    13.36ms
...h.frame_to_csv_mixed.time_frame_to_csv_mixed   218.47ms
...m.time_read_csv_infer_datetime_format_custom    15.82ms
....time_read_csv_infer_datetime_format_iso8601     2.64ms
..._ymd.time_read_csv_infer_datetime_format_ymd     2.92ms
...nch.read_csv_skiprows.time_read_csv_skiprows    16.74ms
...nch.read_csv_standard.time_read_csv_standard    14.00ms
..._dates_iso8601.time_read_parse_dates_iso8601     1.93ms
...h.write_csv_standard.time_write_csv_standard    50.49ms
For Master
=========
io_bench.frame_to_csv.time_frame_to_csv           195.29ms
io_bench.frame_to_csv2.time_frame_to_csv2         303.08ms
...formatting.time_frame_to_csv_date_formatting    12.99ms
...h.frame_to_csv_mixed.time_frame_to_csv_mixed   224.39ms
...m.time_read_csv_infer_datetime_format_custom    15.53ms
....time_read_csv_infer_datetime_format_iso8601     2.64ms
..._ymd.time_read_csv_infer_datetime_format_ymd     2.80ms
...nch.read_csv_skiprows.time_read_csv_skiprows    16.04ms
...nch.read_csv_standard.time_read_csv_standard    13.23ms
..._dates_iso8601.time_read_parse_dates_iso8601     1.78ms
...h.write_csv_standard.time_write_csv_standard    51.60ms

**Comment by gfyoung** [comment]
@jreback : Travis is giving the green light, and I don't see any major timing discrepancies.  Ready to merge if there is nothing else.

**Comment by jreback** [comment]
thanks @gfyoung didn't realize how much duplicative code there was already in tokenizer.c sheesh 👍

**Comment by jreback** [comment]
as an aside, adding ``delim_whitespace`` option to python parser is trivial.

**Pull request closed without merging**